calcite/src/main/java/org/apache/iotdb/calcite/IoTDBFilter.java

```diff
@@ -0,0 +1,321 @@
+    try {
+      return new IoTDBFilter(getCluster(), traitSet, input, condition);
+    } catch (LogicalOptimizeException e) {
+      throw new AssertionError(e.getMessage());
+    }
+  }
+
+  @Override
+  public void implement(Implementor implementor) {
+    implementor.visitChild(0, getInput());
+    implementor.add(deviceToFilterMap, predicates);
+  }
+
+  /**
+   * Translates {@link RexNode} expressions into IoTDB filter operator.
+   */
+  private FilterOperator getIoTDBOperator(RexNode filter) {
+    switch (filter.getKind()) {
+      case EQUALS:
+        return getBasicOperator(SQLConstant.EQUAL, (RexCall) filter);
+      case NOT_EQUALS:
+        return getBasicOperator(SQLConstant.NOTEQUAL, (RexCall) filter);
+      case GREATER_THAN:
+        return getBasicOperator(SQLConstant.GREATERTHAN, (RexCall) filter);
+      case GREATER_THAN_OR_EQUAL:
+        return getBasicOperator(SQLConstant.GREATERTHANOREQUALTO, (RexCall) filter);
+      case LESS_THAN:
+        return getBasicOperator(SQLConstant.LESSTHAN, (RexCall) filter);
+      case LESS_THAN_OR_EQUAL:
+        return getBasicOperator(SQLConstant.LESSTHANOREQUALTO, (RexCall) filter);
+      case AND:
+        return getBinaryOperator(SQLConstant.KW_AND, ((RexCall) filter).getOperands());
+      case OR:
+        return getBinaryOperator(SQLConstant.KW_OR, ((RexCall) filter).getOperands());
+      default:
+        throw new AssertionError("cannot get IoTDBOperator from " + filter);
+    }
+  }
+
+  private FilterOperator getBasicOperator(int tokenIntType, RexCall call) {
+    final RexNode left = call.operands.get(0);
+    final RexNode right = call.operands.get(1);
+
+    FilterOperator operator = getBasicOperator2(tokenIntType, left, right);
+    if (operator != null) {
+      return operator;
+    }
+    operator = getBasicOperator2(tokenIntType, right, left);
+    if (operator != null) {
+      return operator;
+    }
+    throw new AssertionError("cannot translate basic operator: " + call);
+  }
+
+  private FilterOperator getBasicOperator2(int tokenIntType, RexNode left, RexNode right) {
+    switch (right.getKind()) {
+      case LITERAL:
+        break;
+      default:
+        return null;
+    }
+    final RexLiteral rightLiteral = (RexLiteral) right;
+    switch (left.getKind()) {
+      case INPUT_REF:
+        String name = fieldNames.get(((RexInputRef) left).getIndex());
+        return new BasicFunctionOperator(tokenIntType, new Path(name), literalValue(rightLiteral));
+      case CAST:
+        return getBasicOperator2(tokenIntType, ((RexCall) left).getOperands().get(0), right);
+      default:
+        return null;
+    }
+
+  }
+
+  private FilterOperator getBinaryOperator(int tokenIntType, List<RexNode> operands) {
+    FilterOperator filterBinaryTree = new FilterOperator(tokenIntType);
+    FilterOperator currentNode = filterBinaryTree;
+    for (int i = 0; i < operands.size(); i++) {
```

Comment:
better to add a local variance `size = operands.size()` to avoid frequently called in loop

Reply:
Fixed.